_std/macros/atmos.dm

```diff
@@ -334,3 +334,24 @@ proc/gas_text_color(gas_id)
 
 #define _LIST_CONCENTRATION_REPORT(GAS, _, NAME, MIXTURE, LIST) LIST += "[NAME]: [round(MIXTURE.GAS / total_moles * 100)]%";
 #define LIST_CONCENTRATION_REPORT(MIXTURE, LIST) APPLY_TO_GASES(_LIST_CONCENTRATION_REPORT, MIXTURE, LIST)
+
+//Possible states are exposed and intact. sizes are short, medium and long.
+#define SET_PIPE_UNDERLAY(NODE, DIR, SIZE, COLOUR, HIDDEN) do { \
+	if (UNLINT(HIDDEN)) { \
+		src.UpdateOverlays(null, "[DIR]"); \
+		break; \
+		}  \
+	var/pipe_state = NODE ? "intact" : "exposed"; \
+	var/pipe_cached = pipe_underlay_cache["[pipe_state]_[DIR]_[SIZE]"]; \
+	if (!pipe_cached) { \
+		pipe_underlay_cache["[pipe_state]_[DIR]_[SIZE]"] = icon('icons/obj/atmospherics/pipes/pipe_underlays.dmi', "[pipe_state]_[NODE ? null : SIZE]", DIR); \
+		pipe_cached = pipe_underlay_cache["[pipe_state]_[DIR]_[SIZE]"]; \
+		} \
+	var/image/pipe_image = mutable_appearance(pipe_cached); \
```

Comment:
Is caching and generating a mutable appearance from the cache every time actually any better than just making a new image every time?

Reply:
best case, yes. did this with each proc doing it 10000 times.
![image](https://github.com/goonstation/goonstation/assets/62359847/d7f12bd1-8fe8-4bdd-86aa-b59df03b9030)